.gitlab-ci.yml

```diff
@@ -1,14 +1,131 @@
+  - deploy
+
+.fedora_install_deps: &fedora_install_deps
+  before_script:
+    - dnf install -y --refresh python3 python3-devel 'python3dist(pip)'
+    # requirements
+    - dnf install -y 'python3dist(docopt)' 'python3dist(lxml)' 'python3dist(pyyaml)' 'python3dist(pyxattr)' 'python3dist(six)' 'python3dist(future)' 'python3dist(requests)'
+    # dev-requirements
+    - dnf install -y 'python3dist(pytest)' 'python3dist(pytest-xdist)' 'python3dist(pytest-cov)' 'python3dist(mock)' 'python3dist(flake8)' 'python3dist(sphinx)' python3-sphinx_rtd_theme 'python3dist(sphinxcontrib-spelling)' 'python3dist(pyenchant)' 'python3dist(tox)' make gcc which xz xorriso
+
+.distro_test: &distro_test
+  script:
+    - python3 setup.py build
+    - export PYTHONPATH=$(pwd)
+    - cd test/unit/ && pytest-3.7 --no-cov-on-fail --cov=kiwi --cov-report=term-missing --cov-fail-under=100 --cov-config .coveragerc -n $(nproc)
+
+.tox_3_test: &tox_3_test
+  script:
+    - $PYTHON -m venv .env3
+    - source .env3/bin/activate
+    - pip install -r .virtualenv.dev-requirements.txt
+    - $PYTHON setup.py develop
+    - tox "-n $(nproc)"
+
+tox_37_check:
+  stage: test
+  image: opensuse/tumbleweed
+  variables:
+    TOXENV: 'unit_py3_7,check'
+    PYTHON: 'python3'
+  before_script:
+    - zypper --non-interactive refresh
+    - zypper --non-interactive install libxml2-devel libxslt-devel gcc python3 python3-devel python3-pip enchant ShellCheck which mkisofs xz gzip
+  <<: *tox_3_test
+
+tox_36_check:
+  stage: test
+  image: fedora:29
+  variables:
+    TOXENV: 'unit_py3_6'
+    PYTHON: 'python3.6'
+  before_script:
+    - dnf install -y --refresh python36 tox libxml2-devel libxslt-devel gcc enchant make gcc which xz xorriso
+  <<: *tox_3_test
+
+tox_27_check:
+  stage: test
+  image: opensuse/tumbleweed
+  before_script:
+    - zypper --non-interactive refresh
+    - zypper --non-interactive install libxml2-devel libxslt-devel gcc python2 python2-devel python2-virtualenv python2-pip enchant which mkisofs xz gzip
+  script:
+    - virtualenv .env2
+    - source .env2/bin/activate
+    - pip install -r .virtualenv.dev-requirements.txt
+    - python2 setup.py develop
+    - tox -e unit_py2_7 "-n $(nproc)"
+
+OpenSUSE:
+  stage: test
+  image: opensuse/tumbleweed
+  before_script:
+    - zypper --non-interactive refresh
+    - zypper --non-interactive install python3 python3-devel python3-pip
+    # requirements
+    - zypper --non-interactive install python3-docopt python3-lxml python3-PyYAML python3-xattr python3-six python3-future python3-requests
+    # dev-requirements
+    - zypper --non-interactive install python3-pytest python3-pytest-cov python3-mock python3-flake8 python3-Sphinx python3-sphinx_rtd_theme python3-pyenchant python3-tox make gcc which mkisofs xz gzip
+    # pytest-xdist from the repos is broken, use the one from PyPI instead
+    - pip3 install sphinxcontrib-spelling pytest-xdist
+  <<: *distro_test
+
+Fedora:
+  stage: test
+  image: fedora:29
```

Comment:
Do we want to do `fedora:29` or `fedora:latest`? We are using `opensuse/tumbleweed` for the openSUSE jobs, so we should do the same on the Fedora side...

Reply:
Well, Tumbleweed is unversioned by its nature. Although I have to admit that I added the 29 because I though that `latest` would be rawhide (now that I checked I know that is not true).

Anyway, my thoughts behind this are that the CI should be somewhat reproducible and pulling in the latest image means that on every Fedora switch it could break (because package names change, dependencies won't resolve, etc). Using a fixed version means that we consciously upgrade the CI and it doesn't happen accidentally/automatically.

Reply:
Then we should probably be using `opensuse/leap:15.0` then?

Reply:
Yeah, according to my logic we should, but Leap is really old... I'll use `:latest` then.